Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp

```diff
@@ -333,197 +314,94 @@ void Recorder::drawPattern(ImageBuffer& imageBuffer, const FloatRect& destRect,
     recordDrawPattern(imageBuffer.renderingResourceIdentifier(), destRect, tileRect, patternTransform, phase, spacing, options);
 }
 
-void Recorder::save(GraphicsContextState::Purpose purpose)
+void Recorder::updateStateForSave(GraphicsContextState::Purpose purpose)
 {
     ASSERT(purpose == GraphicsContextState::Purpose::SaveRestore);
-
     appendStateChangeItemIfNecessary();
     GraphicsContext::save(purpose);
-    recordSave();
     m_stateStack.append(m_stateStack.last());
 }
 
-void Recorder::restore(GraphicsContextState::Purpose purpose)
+bool Recorder::updateStateForRestore(GraphicsContextState::Purpose purpose)
 {
     ASSERT(purpose == GraphicsContextState::Purpose::SaveRestore);
-
     appendStateChangeItemIfNecessary();
     GraphicsContext::restore(purpose);
 
     if (!m_stateStack.size())
-        return;
+        return false;
 
     m_stateStack.removeLast();
-    recordRestore();
+    return true;
 }
 
-void Recorder::translate(float x, float y)
+bool Recorder::updateStateForTranslate(float x, float y)
 {
     if (!x && !y)
-        return;
+        return false;
     currentState().translate(x, y);
-    recordTranslate(x, y);
+    return true;
 }
 
-void Recorder::rotate(float angleInRadians)
+bool Recorder::updateStateForRotate(float angleInRadians)
 {
     if (WTF::areEssentiallyEqual(0.f, fmodf(angleInRadians, piFloat * 2.f)))
-        return;
+        return false;
     currentState().rotate(angleInRadians);
-    recordRotate(angleInRadians);
+    return true;
 }
 
-void Recorder::scale(const FloatSize& size)
+bool Recorder::updateStateForScale(const FloatSize& size)
 {
     if (areEssentiallyEqual(size, FloatSize { 1.f, 1.f }))
-        return;
+        return false;
     currentState().scale(size);
-    recordScale(size);
+    return true;
 }
 
-void Recorder::concatCTM(const AffineTransform& transform)
+bool Recorder::updateStateForConcatCTM(const AffineTransform& transform)
 {
     if (transform.isIdentity())
-        return;
+        return false;
 
     currentState().concatCTM(transform);
-    recordConcatenateCTM(transform);
+    return true;
 }
 
-void Recorder::setCTM(const AffineTransform& transform)
+void Recorder::updateStateForSetCTM(const AffineTransform& transform)
 {
     currentState().setCTM(transform);
-    recordSetCTM(transform);
 }
 
 AffineTransform Recorder::getCTM(GraphicsContext::IncludeDeviceScale) const
 {
     return currentState().ctm;
 }
 
-void Recorder::beginTransparencyLayer(float opacity)
+void Recorder::updateStateForBeginTransparencyLayer(float opacity)
```

Comment:
The intention is:
- Each drawing command that modifies state should call `Recorder::updateStateForABCD()`.
- Recoder is free to call whatever it needs to implement such state modifications to the GraphicsContext state tracking.

It is infact a GraphicsContext base class bug that the subclasses must call the vfunc base implementations to modify the state.

It would look odd that all state modification functions would call `updateStateForABCD()` but then save and `beginTransparencyLayer()` would call  `Recorder::save()` and `Recorder::beginTransparencyLayer()`. You can also note that the return types and parameters are different between `Recorder::ABCD()` and  `Recorder::updateStateForABCD()`. The intention is to not call the base class version, rather call the state update function that is implemented in whatever way is needed..